### ISSUE TITLE
Integration of RFC2 work with existing PROJ code: axis order and unit issues

### ISSUE THREAD

**Issue by rouault** [OSGeo/PROJ#1178]
(as proj@lists.maptools.org seems to be unable to deliver email right now...)

Now that RFC2 implementation has been merged, come the next steps. One of 
them will be interation in GDAL, and then libgeotiff, so there will be fixes 
and complementary work that will be done as a result of those integrations.

For PROJ itself, one of the main thing I'd like to address is the fact that we 
have now two CRS databases:
- the old one in the data/epsg and data/IGNF files
- the new one in the SQLite3 proj.db database

The new one contains the content of the old one (with some differences, like 
the EPSG database in proj.db is fresher than in the old one). So the plan is 
to kill the old 'epsg' and 'IGNF' file that are used by the +init=file:code 
syntax and go instead into the database to fetch their definitions. If no 
match is found, so for other authorities than EPSG, IGNF or ESRI currently, 
then the existing text file based mechanism would still be used.

All good ? Well, not quite. One of the main issue I foresee before doing any 
implementation is axis order. In the old epsg file, geographic CRS use the 
longitude,latitude order, whereas proj.db and the new code I've developped 
respect the axis order declared by the authority, so in the case of EPSG, 
latitude,longitude for geographic CRS. This also applies for a few projected 
CRS: 'epsg' file uses generally easting,northing order (except for a few 
Transverse Mercator South Orientated projected CRS where it uses the expected 
westing,southing order), whereas EPSG might have northing,easting order for a 
few ones, or southing,westing for a few Krovak-based CRS.

What to do ? I can see different options:

- just use the 'correct' axis order, and mark this as a breaking change. That 
is if you use +init=epsg:4326, you'd have to provide / will get latitude first

- emulate the previous axis order in an attempt of being backward compatible. 
Not necessarily easy to do since that means we have to remember somewhere the 
old axis order and add the inversion when needed in some glue code between old 
and new code. And that would undermine one of the interest of the new work to 
be more axis aware than previously.

- have some switch in proj utilies ('proj', 'cs2cs', 'cct') to specify the 
input and output order (like -input_axis=long,lat -output_axis=east,north) so 
that they can adapt user input/output to the expected axis of the CRS used.
That would still be a breaking change since existing scripts would have to add 
the new switches to preserve existing behaviour.

The same question also arises at the API level. For example,
proj_create_crs_to_crs(ctxt, "epsg:4326", "epsg:32632") currently expects the 
input to be in longitude, latitude order.

There is also the question of units. Traditionnaly, in code paths that would 
expect angular values, PROJ expects radians (frontend utilities do the 
conversion into/from degrees). The new code assume sthe unit to be the one 
declared by the coordinate system in the geographic CRS definition, ie degrees 
in 99% of the case (or grad for a couple odd & old systems like "NTF (Paris)" 
/ EPSG:4807)

Let's consider a practical example. "projinfo -s EPSG:4326 -t EPSG:4807" 
outputs a transformation pipeline which is
"
+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert 
+xy_in=deg +xy_out=rad +step +inv +proj=hgridshift +grids=ntf_r93.gsb +step 
+proj=longlat +ellps=clrk80ign +pm=paris +step +proj=unitconvert +xy_in=rad 
+xy_out=grad +step +proj=axisswap +order=2,1"

The traduction in plain English of this is:
- take the input coordinate and reverse the first 2 compoments
- convert from degree to radian
- do a horizontal grid shift transformation
- apply a prime meridian shift
- convert from radian to grad
- take the ouput coordinate and reverse the first 2 compoments

That is it expects the input to be lat,long in degrees and output lat,long in 
grads, as in the offical definition of those CRS.

Should the PJ* object returned by proj_create_crs_to_crs(ctxt, "epsg:4326", 
"epsg:4807") just use that pipeline, and consequently expect the coordinates 
passed into and received from proj_trans() to also use those orders and units.

Thoughts ?

Even

PS: the database might also contain geographic CRS definitions that are 
longitude,latitude by design. For example I've added an entry for the 
OGC:CRS84, and if you do "projinfo -s OGC:CRS84 -t EPSG:4807", then the 
pipeline doesn't contain the initial axisswap operations.

**Comment by rouault** [comment]
@kbevers @busstoptaktik @hobu I'd appreciate your thoughts on this. Note Martin's proposal in http://lists.maptools.org/pipermail/proj/2018-November/008509.html

**Comment by kbevers** [comment]
Yes, of course. It is on my todo list. I just need some quiet time to gather my thoughts on write them down properly. I'll get around to that tonight.

**Comment by hobu** [comment]
/me looks for the emoji to put his head into the sand

How was a definition supposed to ever know if it was using old or new behavior in the past? If it was GDAL/PROJ -based it was old? 

I think we should refuse the temptation to guess in this situation. Maybe we should be do something obnoxious like GEOS' C++ API and make them set some kind of per-environment `PROJ_USE_BROKEN_AXIS_ORDERING=TRUE` and `PROJ_USE_BROKEN_UNITS=TRUE` variables that makes it very noisy and explicit. If you want to do something that's dirty and polluting, everyone needs to know about it. Silently attempting to do the "right" thing here is going to make it difficult for us to ditch the bad behavior in the long term.

Of course, this pushes the problem up the stack to applications like GDAL, but maybe it more properly belongs up there in the first place. I think GDAL and any other users dependent upon the broken axis ordering are going to need tools to manage their transition to the new codebase at their own pace. I'm not sure what exactly those tools are, and they may have to be equivalent to 1860s military field hospital surgical instruments.

**Comment by kbevers** [comment]
This is a tough one. Let's start with the easy decision: I think that `proj`, `cs2cs` and `cct` should always respect the axis order of both in- and output CRS. 
How that should be handled API-wise is more complicated. I have two suggestions:

#### 1. Always respect CRS axis order

Consider the following example code:

```
PJ_COORD c = {12.0, 56.0, 100, 2018.0};
PJ *P = proj_create_crs_to_crs(ctxt, "epsg:4326", "epsg:4807");
c = proj_reorder_axis_in(P, c, PJ_EAST, PJ_NORTH, PJ_UP, PJ_TIME);
/* here c == {56.0, 12.0, 100, 2018.0} */
c = proj_trans(P, c);
/* here c == {56.0, 12.0, 100, 2018.0} */
c = proj_reorder_axis_out(P, c, PJ_EAST, PJ_NORTH, PJ_UP, PJ_TIME);
/* here c == {12.0, 56.0, 100, 2018.0} */
```

Here the user of the API can use whatever axis ordering they want as long as they know what it is. The `proj_reorder_axis_*` functions then deal with handling the axis reordering of both input and output coordinates.

The upside here is that if you want to be true to the definition of the CRS's then you can skip the `proj_reorder_axis_*` functions. If you have older code that rely on the hard-coded axis order you just wrap your calls to `proj_trans()` in two functions.

The downside is that `proj_trans_array()` and `proj_trans_generic()` doesn't really fit the scheme above.

#### 2. Always expect (easting, northing, up, time) in/out axis ordering

Alternatively we can keep the current setup that always expect the east/west component of a coordinate first, then the north/south component and finally the up/down and time components. We could then introduce similar functions as described above, but this time focusing on presenting the coordinate correctly for the user while keeping the internal structure the same:

```
PJ_COORD c = {56.0, 12.0, 100, 2018.0};
PJ *P = proj_create_crs_to_crs(ctxt, "epsg:4326", "epsg:4807");
c = proj_reorder_axis_in(P, c); 
/* here c == {12.0, 56.0, 100, 2018.0} */
c = proj_trans(P, c);
/* here c == {12.0, 56.0, 100, 2018.0} */
c = proj_reorder_axis_out(P, c);
/* here c == {56.0, 12.0, 100, 2018.0} */
```

This solution will be simpler to adopt for existing API users, since they don't really have to change anything. Users that wan't to respect the correct axis orders can do so by adding the reordering functions.


I lean towards the second solution as I can see it being quite complicated to handle the differing axis orders that will come into play using the first solution. For instance the grid handling code will likely have to be changed somewhat to accomodate different axis order than (east, north, up). My proposed solutions above using functions to reorder the content of `PJ_COORD`'s may not be the best way forward but hopefully they illustrate my point. The same can be done by setting the axis order directly on the `PJ` object. This may be smarter actually.

**Comment by busstoptaktik** [comment]
I do not entirely agree with anyone so far. I'm working on a longer reply, but it will necessarily become rather long. For now it will suffice to say that I believe the filters `proj` and `cs2cs` should not change behaviour, while `cct` probably should. Also: Note the '-r' and '-s' flags to proj and cs2cs, for switching input, resp. output, order.

**Comment by rouault** [comment]
Folks, please review
https://docs.google.com/document/d/1R73_MK6_XFB9k_SLT57LJM8Z-yw8FgOmvhDERwByYeA/edit?usp=sharing
I've added comment rights to PSC members. Other interested can let their comment here, or ask me to give them right

**Comment by busstoptaktik** [comment]
To me, the question of how (not) to handle axis order seems closely related to the question of how (not) to handle the towgs84-mechanism

The WKT1 style “towgs84=...” datum handling promoted a world view that was “good enough for visualization”, by assigning an unspecified version of WGS84 the **magic property** of being “the one true reference frame, cast in concrete and infinitely valid”.

While useful for limited periods of time and limited accuracy requirements, this, of course, does not reflect the geodetic realities. But it allows for careless transformation between any two CRS, by using a weakly defined common pivot system.

Getting more or less rid of towgs84 has taken more than a decade...

The classic cs2cs way of handling axes was the declarative `+axis=...` option which, much like `towgs84=...`, lives under the happy misconception that “somehow an authoritative, eternal, all-complete, all-perfect set of coordinate axes must exist” and additionally: “Flipping and swapping must be a sufficient set of operations to magically turn any other system into the perfect, eternal system”.

Presumably (but not explicitly stated anywhere obvious), the perfect system is a close cousin to a right handed orthonormal system, with a right-pointing axis called “easting” and a perpendicular axis called “northing”. And actually, for most projected coordinate systems, the magic set of operations are sufficient to get a reasonable approximation to the one described.

But what about oblique projections, what about skew coordinate systems, and what about curvilinear geographical coordinates, or geographical coordinates defined in terms of the colatitude? Some of these are probably not widely used, but they exist, and cannot necessarily be converted in simple ways to anything resembling easting-northing.

**So designing a system based on the assumption that swapping and flipping will be sufficient to transform any kind of coordinate system into a-kind-of-scaled-utm is probably as fruitful as designing one based on the assumption that any two datums can be adequately related through a set of two Helmert transformations.**

Using “enu” style labels to understand axes, and calling anything else than “enu” (“nue”, “esd” etc.) an “unusual axis order” is probably more confusing than helpful in the long term since, although a large number of CRS make ample sense when viewed through a “enu” lens, another group do not.

About the only thing we can safely assume about a CRS is that it consists of two or more axes, that their order matters, and that their meaning is a matter for the CRS - to interpret them correctly, we must somehow be able to query the CRS definition.

So many words just to warn that: Please, let us not paint ourselves into a towgs84-style corner regarding the axis handling. It is a non-trivial problem. Please do not jump to conclusions too quickly.

### proj, cs2cs and cct

Now, the `proj` program is 2D, and (if I remember correctly a recent remark by @kbevers) better limited to reading 2D geographical coordinates, and projecting them to 2D projected coordinates. It does not necessarily relate to the datum of a CRS - the projections do not care about datum. And `proj` has always believed that coordinates should be read in longitude-latitude order. Let it just stay that way, and assume that pipelines are not made for `proj`

cs2cs is different - it can be dramatically reduced by letting @rouault’s new material in concert with proj_create_crs_to_crs() to do the hard work.
Which also means we can get rid of all the cruft that was introduced into pj_fwd and pj_inv in order to make them handle the material in the EPSG and IGNF files.

Still, most things working with cs2cs previously should also be able to work today, since (essentially) the background material is identical between proj.db and the EPSG/IGNF combo. cs2cs should be able to find some kind of transformation, although not necessarily the geodetically best.

Finally, cct is the “pipeline processor”, where the system will not try to second-guess what you mean: Arguments are not source and target CRS, but simply the EPSG transformation ID, or whatever transformation pipeline the user has dreamt up. No problems trying to decide automatically which of many possible transformations to select. The user has selected, and lives with the results.

**Comment by rouault** [comment]
> So many words just to warn that: Please, let us not paint ourselves into a towgs84-style corner regarding the axis handling. It is a non-trivial problem

I agree. That said, resistance to change is huge, and a lot (in numbers) of users of PROJ (the ones you use it without knowing they use it) are probably happy enough with 1 to 10 meter accuracy in transforms and the past deviations to the authoritative axis definition. Hence my attempt to try to propose some way of smoothing the transition.

> Let it just stay that way, and assume that pipelines are not made for `proj`

Sure, but it does support +init substitution: `echo "0 0" | PROJ_LIB=data src/proj +init=epsg:32631`

> cs2cs is different - it can be dramatically reduced by letting @rouault’s new material in concert with proj_create_crs_to_crs() to do the hard work.

Yes, that's true. I was considering that a bit. I didn't propose it since making it use the new transformation logic would affect the result people get, but I'm not opposed to it since anyway the expansion of +init= stuff that used to contain towgs84 will be different now

> cct is the “pipeline processor”

Ok, so CRS to CRS style transformations should go only to cs2cs ? We would need to upgrade it to support 4D coordinates.

**Comment by kbevers** [comment]
> Sure, but it does support +init substitution: echo "0 0" | PROJ_LIB=data src/proj +init=epsg:32631

But only partially since `+towgs84`, `+datum` and `+axis` (maybe more?) are not considered. I think it is worth considering disallowing the use of `+init` with `proj`, the benefit being that we emphasise that no datums are involved when projecting geodetic coordinates.

> > cs2cs is different - it can be dramatically reduced by letting @rouault’s new material in concert with proj_create_crs_to_crs() to do the hard work.
> 
> Yes, that's true. I was considering that a bit. I didn't propose it since making it use the new transformation logic would affect the result people get, but I'm not opposed to it since anyway the expansion of +init= stuff that used to contain towgs84 will be different now
> 
>
> > cct is the “pipeline processor”
> 
> Ok, so CRS to CRS style transformations should go only to cs2cs ? We would need to upgrade it to support 4D coordinates.

I'd be super happy if we just ditched `+towgs84` and friends altogether but that is probably still too early. I am in favor of using `cs2cs` as a frontend for `proj_create_crs_to_crs()`. I guess we can add some conditional logic that uses `pj_transform()` when `+towgs84` and friends are specified by the user. At least in a transition period until `proj_api.h` is removed from PROJ. 

It will require a complete rewrite of `cs2cs` but that is in the pipeline no matter what (necessary when `proj_api.h` is removed). Adding 4D support to `cs2cs` is a minor implementation detail in that regard :-)

**Comment by rouault** [comment]
Closing since resolved by #1182
